README
apache-suexecuser for Debian
-----------------

build with dpkg-buildpackage
<possible notes regarding this package - if none, delete this file>

cd c
cp generators/mod_cgi.h generators/mod_cgi.c .
mv mod_cgisuexecuser.c mod_cgisuexecuser.c.oldversion
cp mod_cgi.c mod_cgisuexecuser.c
patch -p0 < mod_cgisuexecuser.c.diff

 -- William Herrin <herrin@dirtside.com>  Mon, 07 Jan 2013 11:15:38 -0500


debian/changelog
apache-suexecuser (2.4.56) unstable; urgency=medium

  * updated to apache 2.4.56 

 -- William Herrin <herrin@dirtside.com>  Sat, 25 Mar 2023 10:26:47 -0700

apache-suexecuser (1.1+2) unstable; urgency=medium

  * remove check for CGI program being in the same directory tree as the
  * current directory. Let the user decide if that's OK in the
  * .apache-suid file

 -- William Herrin <herrin@dirtside.com>  Sat, 25 Mar 2023 10:14:23 -0700

apache-suexecuser (1.1+1) unstable; urgency=medium

  * figuring out version numbering

 -- William Herrin <herrin@cat.dirtside.com>  Wed, 05 Oct 2016 23:24:04 -0400

apache-suexecuser (1.1) unstable; urgency=medium

  *  Apache 2.4.10
  * 

 -- William Herrin <herrin@dirtside.com>  Wed, 01 Jun 2016 13:53:40 -0400

apache-suexecuser (1.0) unstable; urgency=low

  * notes

 -- William Herrin <herrin@dirtside.com>  Mon, 15 Jul 2013 20:19:41 -0400



c/suexec.h
/*
 * suexec.h -- user-definable variables for the suexec wrapper code.
 *             (See README.configure on how to customize these variables.)
 */


#ifndef _SUEXEC_H
#define _SUEXEC_H

/*
 * HTTPD_USER -- Define as the username under which Apache normally
 *               runs.  This is the only user allowed to execute
 *               this program.
 * note: will use unix file system permissions instead
 */

/*
 * UID_MIN -- Define this as the lowest UID allowed to be a target user
 *            for suEXEC.  For most systems, 500 or 100 is common.
 */
#ifndef UID_MIN
#define UID_MIN 100
#endif

/*
 * GID_MIN -- Define this as the lowest GID allowed to be a target group
 *            for suEXEC.  For most systems, 100 is common.
 */
#ifndef GID_MIN
#define GID_MIN 100
#endif

/*
 * USERDIR_SUFFIX -- Define to be the subdirectory under users' 
 *                   home directories where suEXEC access should
 *                   be allowed.  All executables under this directory
 *                   will be executable by suEXEC as the user so 
 *                   they should be "safe" programs.  If you are 
 *                   using a "simple" UserDir directive (ie. one 
 *                   without a "*" in it) this should be set to 
 *                   the same value.  suEXEC will not work properly
 *                   in cases where the UserDir directive points to 
 *                   a location that is not the same as the user's
 *                   home directory as referenced in the passwd file.
 *
 *                   If you have VirtualHosts with a different
 *                   UserDir for each, you will need to define them to
 *                   all reside in one parent directory; then name that
 *                   parent directory here.  IF THIS IS NOT DEFINED
 *                   PROPERLY, ~USERDIR CGI REQUESTS WILL NOT WORK!
 *                   See the suEXEC documentation for more detailed
 *                   information.
 *
 */
#ifndef USERDIR_SUFFIX
#define USERDIR_SUFFIX "www"
#endif
#ifndef WRAPPER_FILE
#define WRAPPER_FILE ".apache-suid"
#endif
#ifndef DOC_ROOT
#define DOC_ROOT "/var/www"
#endif

/*
 * SAFE_PATH -- Define a safe PATH environment to pass to CGI executables.
 *
 */
#ifndef SAFE_PATH
#define SAFE_PATH "/usr/sbin:/usr/bin:/bin"
#endif

#endif /* _SUEXEC_H */


c/suexec-user.c
/*
 * suexec.c -- "Wrapper" support program for suEXEC behaviour for Apache
 *
 * User version. Each user who is allowed to run CGI programs should have
 * a copy of this in the suexecdir which is setuid and setgid to that user's
 * uid and group.
 *
 * Copyright 2004 CrossLink Internet Services
 * Written by William D. Herrin, herrin@dirtside.com
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <errno.h>
#include <regex.h>

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <stdarg.h>

#include "suexec.h"
#include "suexec-util.h"
#include "lowercase.h"

#undef PATH_MAX
#define PATH_MAX 8192
#define AP_MAXPATH PATH_MAX
#define AP_ENVBUF 256
#define LOG_EXEC "stderr"

#define SUEXEC_UMASK 0077

extern char **environ;


void fail(const int errornumber, const char *fmt,...)
{
    va_list ap;

    printf ("Content-Type: text/plain\r\n\r\nsuexec error: ");
    va_start (ap, fmt);
    vprintf (fmt, ap);
    va_end (ap);
    printf ("\r\n");
    exit (errornumber);
}

struct passwd *shifttoeuid(char const *cmd, uid_t *uid, gid_t *gid) {
/* Change very completely to the requested UID and return the pw structure
 * to the caller */
  struct passwd *pw;
  
  /* printf ("euid=%i, uid=%i, egid=%i, gid=%i\n",geteuid(),getuid(),
	getegid(),getgid()); */

  *uid = geteuid();

  /*
   * Error out if attempt is made to execute as root or as
   * a UID less than UID_MIN.  Tsk tsk.
   */
  if ((*uid == 0) || (*uid < UID_MIN)) {
	fail(107,"crit: cannot run as forbidden uid (%d/%s)\n", *uid, cmd);
  }


  setreuid (*uid,*uid);

  pw = getpwuid(*uid);
  if (pw==NULL) {
	fail(102,"crit: invalid uid: (%ld)\n", uid);
  }

  *gid = pw->pw_gid;
  /*
   * Error out if attempt is made to execute as root group
   * or as a GID less than GID_MIN.  Tsk tsk.
   */
  if ((*gid == 0) || (*gid < GID_MIN)) {
	fail(108,"crit: cannot run as forbidden gid (%d/%s)\n", *gid, cmd);
  }
  setregid (*gid,*gid);
  /*printf ("euid=%i, uid=%i, egid=%i, gid=%i\n",geteuid(),getuid(),
	getegid(),getgid());*/


  if ((getegid()!=*gid)||(getgid()!=*gid)) {
	fail(109,"emerg: failed to setgid (%ld!=%ld!=%ld: %s)\n", *gid,
		getegid(),getgid(),cmd);
  }

  /* printf ("euid=%i, uid=%i, egid=%i, gid=%i\n",geteuid(),getuid(),
	getegid(),getgid()); */

  return pw;
}

void validate_dirfilestat (
	char const *file
,	struct stat *dir_info
,	struct stat *file_info
,	uid_t uid
,	gid_t gid
,	char const *cwd
,	char const *permfile
) {
    char dn[AP_MAXPATH];	/* dirname */
    int dirlen;

    if ((strlen(file)+2) > AP_MAXPATH) {
	fail(152,"error: very long path\n");
    }
    strcpy (dn,file);
    dirlen = strlen(dn);
    for (dirlen = strlen(dn)-1;  ((dirlen>=0)&&(dn[dirlen]!='/')); dirlen--) {
      dn[dirlen]=0;
    }
    dirlen++;
    if (dirlen<=0) fail(152,"error: no directory in %s\n",file);

    /* fn=file+dirlen;
    while (*fn=='/') fn++; */

    /*
     * Stat the cwd and verify it is a directory, or error out.
     */
    if (((lstat(dn, dir_info)) != 0) || !(S_ISDIR(dir_info->st_mode))) {
	fail(115,"error: cannot stat directory: (%s)\n", dn);
    }

    /*
     * Error out if cwd is writable by others.
     */
    if ((dir_info->st_mode & S_IWOTH) || (dir_info->st_mode & S_IWGRP)) {
	fail(116,"error: directory is writable by others: (%s)\n", dn);
    }

    /*
     * Error out if we cannot stat the program.
     */
    if (lstat(file, file_info) != 0) {
	fail(117,"error: cannot stat file: (%s)\n", file);
    }

    /*
     * Error out if the file is a link and the linked program does not
     * match the security constraints
     */
    if (S_ISLNK(file_info->st_mode)) {
      char buf[AP_MAXPATH],failreason[AP_MAXPATH];

      strncpy (buf,cwd,AP_MAXPATH);
      buf[AP_MAXPATH-1] = 0;
      if (followlinks(buf,AP_MAXPATH,file)==NULL) {
	fail(117,"error: failed to follow links: (%s) cwd=%s\n", file,cwd);
      }

      if (permfile!=NULL) {
	/* Note: assume that the perm files are already checked... */
        if (!read_wrapper_permissions("/etc/apache2/apache-suid",
  	  buf, failreason, AP_MAXPATH)) {
          fail (117,"emerg: wrapper error in /etc/apache2/apache-suid\r\n"
		" on %s\r\n %s\r\n", buf, failreason);
        }
        if (!read_wrapper_permissions(permfile, 
	  buf, failreason, AP_MAXPATH)) {
          fail (117,"emerg: wrapper error in %s\r\n on cgi file %s\r\n %s\r\n",
		permfile, buf, failreason);
        }
      } else {
        fail(117,"error: file is a symbolic link (not allowed): (%s)\n", file);
      }
    }

    /*
     * Error out if the program is writable by others.
     */
    if ( ((file_info->st_mode & S_IWOTH) || (file_info->st_mode & S_IWGRP)) 
	&& (!S_ISLNK(file_info->st_mode)) ) {
	fail(118,"error: file is writable by others: (%s / %s)\n", dn, file);
    }

    /*
     * Error out if the file is setuid or setgid.
     */
    if ((file_info->st_mode & S_ISUID) || (file_info->st_mode & S_ISGID)) {
	fail(119,"error: file is either setuid or setgid: (%s/%s)\n", dn,file);
    }

    /*
     * Error out if the target name/group is different from
     * the name/group of the cwd or the program.
     */
    if ((uid != dir_info->st_uid) ||
	(gid != dir_info->st_gid) ||
	(uid != file_info->st_uid) ||
	(gid != file_info->st_gid)) {
	fail(120,"error: target uid/gid (%ld/%ld) mismatch "
		"with directory (%ld/%ld) or file (%ld/%ld)\n",
		uid, gid,
		dir_info->st_uid, dir_info->st_gid,
		file_info->st_uid, file_info->st_gid);
    }
  return;
}


int main(int argc, char *argv[])
{
    uid_t uid;			/* user information          */
    gid_t gid;			/* target group placeholder  */
    char *target_homedir;	/* target home directory     */
    char *cmd;			/* command to be executed    */
    char failreason[AP_MAXPATH];/* failure reason            */
    char cwd[AP_MAXPATH];	/* current working directory */
    char pfn[AP_MAXPATH];	/* permission file name      */
    struct passwd *pw;		/* password entry holder     */
    struct group *gr;		/* group entry holder        */
    struct stat dir_info;	/* directory info holder     */
    struct stat prg_info;	/* program info holder       */

#ifdef _OSD_POSIX
    char *actual_uname;		/* actual user name          */
    char *actual_gname;		/* actual group name         */
#endif

    cmd = argv[0];

    /* Shift totally to the effective uid, forgetting the old. */
    pw = shifttoeuid(cmd,&uid,&gid);

    /*
     * Save these for later since initgroups will hose the struct
     */
    uid = pw->pw_uid;
    target_homedir = strdup(pw->pw_dir);

#ifdef _OSD_POSIX
    actual_uname = strdup(pw->pw_name);
#endif
    /* Verify that the target program is under the current working
     * directory */
    if (getcwd(cwd, AP_MAXPATH) == NULL) {
        fail(111,"emerg: cannot get current working directory\n");
    }
    /* This fails on the legitimate case of a symbolic link
     * to the directory from elsewhere. Let the user decide.
     */
    /*if (strncmp(cmd,cwd,strlen(cwd))) {*/
      /* trying to run a program in a different directory tree */
    /*  fail(104,"error: invalid path (%s) for command (%s)\n", cwd, cmd);
    }*/
    /*
     * Check for attempts to back up out of the current directory,
     * to protect against attacks.  If any are
     * found, error out.  Naughty naughty crackers.
     */
    if ((cmd[0] != '/') || (strstr(cmd, "/../") != NULL)) {
        fail(104,"error: invalid command (%s)\n", cmd);
    }

    if ((gr = getgrgid(gid)) == NULL) {
	    fail(106,"crit: invalid target group: (%i)\n", (int) gid);
	}

#ifdef _OSD_POSIX
    actual_gname = strdup(gr->gr_name);
    /*
     * Initialize BS2000 user environment
     */
    {
	pid_t pid;
	int status;

	switch (pid = ufork(actual_uname))
	{
	case -1:	/* Error */
	    fail(150,"emerg: failed to setup bs2000 environment for user "
		    "%s: %s\n",
		    actual_uname, strerror(errno));
	case 0:	/* Child */
	    break;
	default:	/* Father */
	    while (pid != waitpid(pid, &status, 0))
		;
	    /* @@@ FIXME: should we deal with STOP signals as well? */
	    if (WIFSIGNALED(status)) {
		kill (getpid(), WTERMSIG(status));
	    }
	    exit(WEXITSTATUS(status));
	}
    }
#endif /* _OSD_POSIX */



    /* User is legit, and we're now running as that user. Now, check out
     * the file we plan to run.
     */

    /*
     * Get the current working directory, as well as the proper
     * document root (dependant upon whether or not it is a
     * ~userdir request).  Error out if we cannot get either one,
     * or if the current working directory is not in the docroot.
     * Use chdir()s and getcwd()s to avoid problems with symlinked
     * directories.  Yuck.
     */
    if (getcwd(cwd, AP_MAXPATH) == NULL) {
	fail(111,"emerg: cannot get current working directory\n");
    }

    validate_dirfilestat ("/etc/apache2/apache-suid",
	 &dir_info,&prg_info,0,0,NULL,NULL);
    if (!read_wrapper_permissions("/etc/apache2/apache-suid",
	cmd, failreason, AP_MAXPATH)) {
      fail (114,"emerg: wrapper error in /etc/apache2/apache-suid\r\n %s",
	failreason);
    }
    snprintf (pfn,AP_MAXPATH,"%s/.apache-suid",target_homedir);
    validate_dirfilestat (pfn,&dir_info,&prg_info,uid,gid,NULL,NULL);
    if (!read_wrapper_permissions(pfn, cmd, failreason, AP_MAXPATH)) {
      fail (114,"emerg: wrapper error in %s\r\n %s", pfn, failreason);
    }

    /*
     * Stat the cwd and verify it is a directory, or error out.
     */
    validate_dirfilestat (cmd,&dir_info,&prg_info,uid,gid,cwd,pfn);

    /*
     * Error out if the program is not executable for the user.
     * Otherwise, she won't find any error in the logs except for
     * "[error] Premature end of script headers: ..."
     */
    if (!(prg_info.st_mode & S_IXUSR)) {
	fail(121,"error: file has no execute permission: (%s/%s)\n", cwd, cmd);
    }

#ifdef SUEXEC_UMASK
    /*
     * umask() uses inverse logic; bits are CLEAR for allowed access.
     */
    /*if ((~SUEXEC_UMASK) & 0022) {
	log_err("notice: SUEXEC_UMASK of %03o allows "
		"write permission to group and/or other\n", SUEXEC_UMASK);
    }*/
    umask(SUEXEC_UMASK);
#endif /* SUEXEC_UMASK */
    environ = clean_env(environ);

    /*
     * Execute the command, replacing our image with its own.
     */
#ifdef NEED_HASHBANG_EMUL
    /* We need the #! emulation when we want to execute scripts */
    {
	extern char **environ;

	ap_execve(cmd, argv, environ);
    }
#else /*NEED_HASHBANG_EMUL*/
    execv(cmd, argv);
#endif /*NEED_HASHBANG_EMUL*/

    /*
     * (I can't help myself...sorry.)
     *
     * Uh oh.  Still here.  Where's the kaboom?  There was supposed to be an
     * EARTH-shattering kaboom!
     *
     * Oh well, log the failure and error out.
     */
    fail(255,"emerg: (%d)%s: exec failed (%s)\n", errno, strerror(errno), cmd);
  return 255;
}


c/suexec-util.c
/* suexec-util.c - utility functions for the suexec programs that are common
 * to all.
 *
 *
 * Copyright 2004 CrossLink Internet Services
 * Written by William D. Herrin, herrin@dirtside.com
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>

#include "suexec-util.h"
#include "lowercase.h"

char *followlinks (
/* Find the real location of the file "startname" by following all
 * symbolic links until you reach the real location */
	char *buf
,	int maxbuf
,	const char *startname
) {
  int i, j, done;
  char bufa[2048], bufb[2048];
  char *p, *q, *remainder;
  int maxdepth = 256;

  maxbuf=(maxbuf<2048)?maxbuf:2048;
  maxbuf--;
  p = buf;
  if (startname[0]!='/') {
    if (!getcwd(buf,maxbuf)) return NULL;
    p = buf + strlen(buf);
  }
  j = strlen(startname);
  if ((p-buf+j+2) > maxbuf) return NULL;
  if (p!=buf) {
    *p='/';
    p++;
  }
  strcpy (p,startname);

  /* Scan each element in buf */
  remainder=buf;
  bufa[0]=0;
  done = 0;
  while (!done) {
    if (maxdepth<0) {
      /* printf ("Oops! maxdepth=%i\n",maxdepth); */
      return NULL; 
    }
    if (*remainder==0) done=1;
    /* printf ("Check: %s + %s (%i)\n",bufa,remainder,errno); */
    i = -1;
    if (bufa[0]!=0) {
      i = readlink (bufa,bufb,maxbuf);
      if ((errno>0)&&(errno!=EINVAL)) {
        /* printf ("Oops! Errno=%i\n",errno); */
        return NULL;
      } 
    }
    if (i<0) { /* Not a link, move to the next element */
      if ((remainder[0]=='/')&&(remainder[1]=='/')) {
        remainder+=1;
	continue;
      }
      if ((remainder[0]=='/')&&(remainder[1]=='.')&&(remainder[2]=='/')) {
        remainder+=2;
	continue;
      }
      if ((remainder[0]=='/')&&(remainder[1]=='.')&&(remainder[2]=='.')
		&&(remainder[3]=='/')) {
        remainder+=3;
	p=bufa+strlen(bufa)-1;
	while ((p>=bufa)&&(*p!='/')) p--;
	while ((p>=bufa)&&(*p=='/')) p--;
	p[1]=0;
	continue;
      }
      for (p=remainder,q=bufa+strlen(bufa); (*p=='/'); p++,q++) *q=*p;
      for (; (*p!=0)&&(*p!='/'); p++,q++) *q=*p;
      *q=0;
      remainder=p; 
      continue;
    }
    bufb[i]=0; 
    maxdepth --;
    if (bufb[0]=='/') { /* replace path elements so far */
      if ((strlen(bufb)+strlen(remainder)+2)>maxbuf) return NULL;
      strcpy (bufb+strlen(bufb),remainder);
      strcpy (buf,bufb);
      for (p=buf,q=bufa; (*p=='/'); p++,q++) *q=*p;
      for (; (*p!=0)&&(*p!='/'); p++,q++) *q=*p;
      *q=0;
      remainder=p; 
      done = 0;
      continue;
    }
    /* not /, so add to the remainder and remove the last path element */
    i = strlen(bufb);
    if ((strlen(buf)+i+strlen(remainder)+3)>maxbuf) return NULL;
    strcpy (bufb+i,remainder);
    buf[0]='/';
    strcpy (buf+1,bufb);
    remainder=buf;
    p=bufa+strlen(bufa)-1;
    while ((p>=bufa)&&(*p!='/')) p--;
    while ((p>=bufa)&&(*p=='/')) p--;
    p[1]=0;
    done = 0;
  }

  strcpy (buf,bufa);
  return buf;
}

/* Main segment to test followlinks... strike
  int main (int argc, char **argv) {
    char buf[2048];
    printf ("Real path: %s\n",followlinks (buf,2048, argv[1]));
  }
*/

double wrappercurrentload = -1.0;

double getload () {
  FILE *f;
  char line[100], *s;;

  if (wrappercurrentload>=0.0) return wrappercurrentload;

  f = fopen ("/proc/loadavg","rt");
  if (!f) return (0.5); /* Return a guess of 0.50 if I don't know! */
  fgets (line,99,f);
  fclose (f);
  for (s=line; (*s=='.') || ((*s>='0')&&(*s<='9')); s++) ;
  *s = 0;
  if ((*line)==0) return (0.4); /* Return a guess of 0.40 if I don't know! */

  wrappercurrentload = strtod (line,NULL);
  return wrappercurrentload;
}

int read_wrapper_permissions (
/* Check the permissions file to determine if the listed program is
 * allowed to run for this user.
 * return 1 on success. Return 0 and set wrapper_fail_reason on failure.
 */
	const char *permissionfilename
,	const char *cgifilename
,	char *wrapper_fail_reason
,	int max_fail_len
) {
  FILE *f;
  struct stat wrapstat;
  char s[2048], *p;
  int vallen, cgifnlen;

  if (((lstat(permissionfilename, &wrapstat)) != 0) 
	|| !(S_ISREG(wrapstat.st_mode))) {
	snprintf (wrapper_fail_reason,max_fail_len,
		"%s is not a plain text file", permissionfilename);
        return 0;
  }

  cgifnlen = strlen(cgifilename);
  f = fopen(permissionfilename,"r");
  if (f==NULL) {
    snprintf (wrapper_fail_reason,max_fail_len,
		"could not open %s", permissionfilename);
    return 0;
  }
  while (!feof(f)) {
    if (!fgets(s,2047,f)) {
      fclose(f);
      snprintf (wrapper_fail_reason,max_fail_len,
		"no explicit authorization in %s before end seeking %s",
		permissionfilename, cgifilename);
      return 0;
    }
    if (s[0]=='#') continue;
    for (p = s + strlen(s)-1; (p>=s)&& 
	((*p=='\r')||(*p=='\n')||(*p==' ')||(*p=='\t')); p--) *p=0;
    for (p=s; (*p!=0)&&(*p!=':'); p++) ;
    if (*p==':') {
      *p=0;
      p++;
      while ((*p==' ')||(*p=='\t')) p++;
    }
    vallen = strlen(p);
    strlwr (s);
    /* printf ("WRAPPER: command %s\n",s); */
    if (!strcmp(s,"run")) {
      if (vallen<1) continue;
      /* printf ("RUN: compare %s to %s\n",p,cgifilename); */
      if (p[vallen-1]!='*') {
	if (vallen!=cgifnlen) continue;
        if (!strcmp(p,cgifilename)) {
          fclose (f);
          return 1;
	}
	continue;
      }
      if (!strnncmp (p,cgifilename,vallen-1)) {
        fclose (f);
        return 1;
      }
      continue;
    }
    if (!strcmp(s,"stop")) {
      if (vallen<1) {
        fclose(f);
	snprintf (wrapper_fail_reason,max_fail_len,"empty stop: line in %s",
		permissionfilename);
        return 0;
      }
      if (p[vallen-1]!='*') {
	if (vallen!=cgifnlen) continue;
        if (!strcmp(p,cgifilename)) {
          fclose (f);
	  snprintf (wrapper_fail_reason,max_fail_len,
		"matched line stop:%s in %s", p,permissionfilename);
          return 0;
	}
	continue;
      }
      if (!strnncmp (p,cgifilename,vallen-1)) {
        fclose (f);
        snprintf (wrapper_fail_reason,max_fail_len,
		"matched line stop:%s in %s", p,permissionfilename);
        return 0;
      }
      continue;
    }
    if (!strcmp(s,"maxload")) {
      /* Check the load average here */
      double load, maxload;

      maxload = strtod(p,NULL);
      load = getload();
      if (load>maxload) {
        snprintf (wrapper_fail_reason,max_fail_len,
		"current load average %f > maximum %f in %s. Try again later.", 
		load, maxload, permissionfilename);
        fclose(f);
        return 0;
      }
      continue;
    }
    snprintf (wrapper_fail_reason,max_fail_len,
		"unknown command '%s' in %s", s, permissionfilename);
    fclose(f);
    return 0;
  }
  fclose(f);
  snprintf (wrapper_fail_reason,max_fail_len,"no explicit authorization in %s",
		permissionfilename);
  return 0;
}

typedef struct {
  char *name;
  int len;
} safelisttype;

safelisttype safe_env_lst[] =
{
    {.name="AUTH_TYPE",.len=-1},
    {.name="CONTENT_LENGTH",.len=-1},
    {.name="CONTENT_TYPE",.len=-1},
    {.name="DATE_GMT",.len=-1},
    {.name="DATE_LOCAL",.len=-1},
    {.name="DOCUMENT_NAME",.len=-1},
    {.name="DOCUMENT_PATH_INFO",.len=-1},
    {.name="DOCUMENT_ROOT",.len=-1},
    {.name="DOCUMENT_URI",.len=-1},
    {.name="FILEPATH_INFO",.len=-1},
    {.name="GATEWAY_INTERFACE",.len=-1},
    {.name="LAST_MODIFIED",.len=-1},
    {.name="PATH_INFO",.len=-1},
    {.name="PATH_TRANSLATED",.len=-1},
    {.name="QUERY_STRING",.len=-1},
    {.name="QUERY_STRING_UNESCAPED",.len=-1},
    {.name="REMOTE_ADDR",.len=-1},
    {.name="REMOTE_HOST",.len=-1},
    {.name="REMOTE_IDENT",.len=-1},
    {.name="REMOTE_PORT",.len=-1},
    {.name="REMOTE_USER",.len=-1},
    {.name="REDIRECT_QUERY_STRING",.len=-1},
    {.name="REDIRECT_STATUS",.len=-1},
    {.name="REDIRECT_URL",.len=-1},
    {.name="REQUEST_METHOD",.len=-1},
    {.name="REQUEST_URI",.len=-1},
    {.name="SCRIPT_FILENAME",.len=-1},
    {.name="SCRIPT_NAME",.len=-1},
    {.name="SCRIPT_URI",.len=-1},
	/* SCRIPT_URI seems to come out wrong, when it exists at all! */
    {.name="SCRIPT_URL",.len=-1},
    {.name="SERVER_ADMIN",.len=-1},
    {.name="SERVER_NAME",.len=-1},
    {.name="SERVER_ADDR",.len=-1},
    {.name="SERVER_PORT",.len=-1},
    {.name="SERVER_PROTOCOL",.len=-1},
    {.name="SERVER_SOFTWARE",.len=-1},
    {.name="UNIQUE_ID",.len=-1},
    {.name="USER_NAME",.len=-1},
    {.name="TZ",.len=-1},
    {.name="REDIRECT_HTTPS",.len=-1},
    {.name="HTTPS",.len=-1},
    {.name=NULL,.len=-1}
};

safelisttype specially_cleaned_env_lst[] =
{
    {.name="DOCUMENT_ROOT",.len=-1},
    {.name=NULL,.len=-1}
};


#ifndef SAFE_PATH
#define SAFE_PATH "/usr/sbin:/usr/bin:/bin"
#endif
#define AP_ENVBUF 256

char **clean_env(char **environ)
{
    char pathbuf[512];
    char **cleanenv;
    char **ep;
    int cidx = 0;
    int idx, httphostlen=0, skipit, specially_cleaned=0;
    const char *scriptfilename=NULL, *docroot=NULL;
    char *httphost=NULL, *s, *p;

    if ((cleanenv = (char **) calloc(AP_ENVBUF, sizeof(char *))) == NULL) {
        printf ("Content-type: text/plain\r\n\r\nsuexec emerg: failed to malloc memory for environment\n");
        exit(120);
    }

    sprintf(pathbuf, "PATH=%s", SAFE_PATH);
    cleanenv[cidx] = strdup(pathbuf);
    cidx++;


    /* Check the vars. */
    for (ep = environ; *ep ; ep++) {
      if (!strnncmp(*ep, "HTTP_HOST=",10)) httphost = strdup((*ep)+10);
      else if (!strnncmp(*ep, "SCRIPT_FILENAME=",16))
                scriptfilename = (*ep)+16;
      else if (!strnncmp(*ep, "DOCUMENT_ROOT=",14)) docroot = *ep;
    }
    if (httphost) {
      for (s = httphost; *s; s++) if (*s==':') *s=0;
      httphostlen = strlen(httphost);
    }
    if (httphost && scriptfilename && 
	((strnncmpi(scriptfilename,"/home/",6)==0) ||
	 (strnncmpi(scriptfilename,"/var/www/virtual/",17)==0) )
	) {
      /* Fake a <Virtual> for these guys! */
      specially_cleaned=1;
      /* set DOCROOT to a reasonable part of SCRIPT_FILENAME */
      s = (char*) malloc (sizeof(char) * (20 + strlen(scriptfilename)));
      sprintf (s,"DOCUMENT_ROOT=%s",scriptfilename);
      if (strnncmpi(s+14,"/home/",6)==0) {
        /* use the part ending with /www/ (user home dir) */
        for (p=s+20; *p && strnncmpi(p,"/www/",5); p++);
        if (*p) p[4]=0;
      } else if (strnncmpi(s+14,"/var/www/virtual/",17)==0) {
	/* /var/www/virtual/name/www.crosslink.net/ */
        /* Use the part ending with the host name */
	if (strnncmpi(s+14+17,"name/",5)==0) {
          for (p=s+14+17+5; *p && strnncmpi(p,httphost,httphostlen); p++);
          if (*p) {
            for (; *p && (*p!='/'); p++) ;
            *p=0;
          }
        } else { /* /var/www/virtual/12/10.11.12.13/ */
          for (p=s+14+17; (*p>='0') && (*p<='9') ; p++);
          if (*p=='/') {
            for (p++; ((*p>='0') && (*p<='9')) || (*p=='.') ; p++);
            if (*p=='/') *p = 0;
          }
        }
      } else { /* Use existing DOCUMENT_ROOT */
        free (s);
        s = NULL;
        if (docroot) s = strdup (docroot);
      }
      if (s) {
        cleanenv[cidx] = s;
        cidx++;
      }

      /* PATH_TRANSLATED comes out wrong! But in old version too. Ignore? */
  
    }

    /* initialize safe_env_list */
    for (idx = 0; safe_env_lst[idx].name; idx++) {
      safe_env_lst[idx].len = strlen(safe_env_lst[idx].name);
    }
    for (idx = 0; specially_cleaned_env_lst[idx].name; idx++) {
      specially_cleaned_env_lst[idx].len = 
	strlen(specially_cleaned_env_lst[idx].name);
    }


    for (ep = environ; *ep && cidx < AP_ENVBUF-1; ep++) {
        if (specially_cleaned) {
          skipit = 0;
          for (idx = 0; specially_cleaned_env_lst[idx].name; idx++) {
            if (!strnncmp(*ep, specially_cleaned_env_lst[idx].name,
                       specially_cleaned_env_lst[idx].len )) {
              skipit = 1;
              break;
            }
          }
	  if (skipit) continue;
        }
        if (!strnncmp(*ep, "HTTP_", 5)) {
            cleanenv[cidx] = *ep;
            cidx++;
        }
        else {
            for (idx = 0; safe_env_lst[idx].name; idx++) {
                if (!strnncmp(*ep, safe_env_lst[idx].name,
                             safe_env_lst[idx].len )) {
                    cleanenv[cidx] = *ep;
                    cidx++;
                    break;
                }
            }
        }
    }

    cleanenv[cidx] = NULL;

    return cleanenv;
}

